### PR TITLE
Fix clojure-find-ns 

### PR DESCRIPTION
**Pull request by p4v4n**
- when ns is preceded by whitespace or inside comment form.
- closes https://github.com/clojure-emacs/clojure-mode/issues/593

### CODE REVIEW

File: clojure-mode.el
Comment (bbatsov) [inline clojure-mode.el:2278]: Perhaps we should add some comments here?

**Comment by vemv** [comment]
Nice work!

The other day I tried to fix a similar issue but my fix broke other things. Could you check if adding these these tests would also pass?

https://github.com/clojure-emacs/clojure-mode/commit/d6a43eef7ca214bdba649d0681b8e8ddc4d9d8f8

I was particularly interested in this case:

```clj
1 (ns foo)
```

however, if it's too hard of a challenge (as it was for me), feel free to leave it.

File: clojure-mode.el
Comment (p4v4n) [reply clojure-mode.el:2279]: The change is to handle whitespace before comment form. Not sure if this is the best way to do it though.
Added a couple of comments now to make the intent more clear.

Comment (bbatsov) [reply clojure-mode.el:2279]: My point was mainly that it looks a bit weird that we're going forward and backwards back-to-back. If that's used often in the code we can make a function I guess.

Comment (p4v4n) [reply clojure-mode.el:2279]: Agreed. But I am not sure what is the correct abstraction anymore as it's more of a hack.

This is the third occurrence in the codebase but there are other places where similar logic was used with different functions and some places where edge cases are not handled. Even my current change will still break if there are multiple expressions in the same line.

Looks like all these extra steps are only needed in many places because of `beginning-of-defun` behaving 'incorrectly'.
For ex:
```
1 (ns abc) (defn abc []
             (comment
               (println "try (clojure-beginning-of-defun-function) from here"))
             "abc")
```
Now running `(clojure-beginning-of-defun-function)` or `(beginning-of-defun)` from any of the inner expressions takes the cursur to the beginning of the line (before 1).
The expected behaviour IMO from the name would be having the cursor before `(defn`.
The docstring mentions this but there is no way to easy way to control the end location of the cursor.

clojure-mode uses beginning-of-defun a lot so there are a lot of unhandled edge cases.
For example take these two code blocks:
```
(defn s1 [x] (str x)) (+ 1 2)

(defn s2
    [x] (str x))
```
```
(defn s1 [x] (str x)) 
(+ 1 2)
(defn s2
    [x] (str x))
```

Run `M-x cider-eval-defun-at-point`(C-c C-c) inside (+ 1 2) and the result is different in the two scenarios.

Comment (p4v4n) [reply clojure-mode.el:2279]: tldr: IIUC bad choice of abstraction from `beginning-of-defun` fn has propagated many edge cases possibly in all lisp modes.
Maybe I am missing something basic or someone already solved this issue.

Comment (p4v4n) [reply clojure-mode.el:2279]: @bbatsov I have raised a separate PR for replacing `beginning-of-defun` fn completely here.
https://github.com/clojure-emacs/clojure-mode/pull/663
After merging that PR these two lines here can be removed.

Comment (bbatsov) [reply clojure-mode.el:2279]: Thanks for diving deep into this!

Comment (p4v4n) [reply clojure-mode.el]: Merged latest master and removed the extra lines here.
Will wait for @vemv to resolve the other conversation.

**Comment by bbatsov** [comment]
> A simpler fix would be take your regex change and delete this unlikely test case! Allowing any forms before comment is far more likely scenario than having fake ns in metadata.

I agree.

**Comment by bbatsov** [comment]
Btw, you'll also have to rebase this on top of `master`.

**Comment by p4v4n** [comment]
> Btw, you'll also have to rebase this on top of `master`.

Already merged latest master. The commits should be squashed here so didn't bother with a rebase.

**Comment by vemv** [comment]
@p4v4n thanks much for looking into that!

I'd be fine with no action for the time being. Perhaps, a year from now most people will be using clojure-ts-mode, which since isn't regex-based, shouldn't be as prone to false positives.

File: test/clojure-mode-util-test.el
Comment (vemv) [inline test/clojure-mode-util-test.el:99]:
```suggestion
      (expect (clojure-find-ns) :to-equal "bar")))
```

File: test/clojure-mode-util-test.el
Comment (vemv) [inline test/clojure-mode-util-test.el:85]: I'd suggest to also include the following case:

```clj
(comment
  (ns foo))
```

...Asserting that `(clojure-find-ns)` is nil.

That would help us prevent anything like an infinite loop.

**Review by vemv — APPROVED** [review]
LGTM after a couple minor suggestions

**Comment by vemv** [comment]
> A simpler fix would be take your regex change and delete this unlikely test case! Allowing any forms before comment is far more likely scenario than having fake ns in metadata.
>> I agree.

When put like this, I actually agree too. (I could as well flip a coin tbh). Feel free to tackle it in a follow-up PR.

**Comment by p4v4n** [comment]
> When put like this, I actually agree too. (I could as well flip a coin tbh). Feel free to tackle it in a follow-up PR.

I was partially joking about deleting the test case. My past experiences with emacs lisp regexes were not very pleasant.(I haven't used the rx macros though TBF)

Will work on it in a separate branch.We can go with the compromise solution if the proper regex is difficult.

File: test/clojure-mode-util-test.el
Comment (p4v4n) [reply test/clojure-mode-util-test.el:85]: Made both of the suggested changes.
Please squash the commits while merging.

**Pull request merged**
